segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -120,295 +139,572 @@
+    }
 
-            // Step 3: Commit externally.
-            // This operation may call external storage.
-            if (!lazyWrite || (bufferedTxnData.size() > maxEntriesInTxnBuffer)) {
-                log.trace("Persisting all modified keys (except pinned)");
-                val toWriteList = modifiedValues.stream().filter(entry -> !entry.isPinned()).collect(Collectors.toList());
-                writeAll(toWriteList);
-                log.trace("Done persisting all modified keys");
-
-                // Mark written keys as persisted.
-                for (val writtenData : toWriteList) {
-                    writtenData.setPersisted(true);
-                }
+    /**
+     * Executes external commit step.
+     */
+    private CompletableFuture<Void> executeExternalCommitAction(MetadataTransaction txn) {
+        // Execute external commit step.
+        try {
+            if (null != txn.getExternalCommitStep()) {
+                txn.getExternalCommitStep().call();
             }
+        } catch (Exception e) {
+            log.error("Exception during execution of external commit step", e);
+            throw new CompletionException(new StorageMetadataException("Exception during execution of external commit step", e));
+        }
+        return CompletableFuture.completedFuture(null);
+    }
 
-            // Execute external commit step.
-            try {
-                if (null != txn.getExternalCommitStep()) {
-                    txn.getExternalCommitStep().call();
-                }
-            } catch (Exception e) {
-                log.error("Exception during execution of external commit step", e);
-                throw new StorageMetadataException("Exception during execution of external commit step", e);
+    private void validateCommit(MetadataTransaction txn, Map<String, TransactionData> txnData, ArrayList<String> modifiedKeys, ArrayList<TransactionData> modifiedValues) {
+        for (val entry : txnData.entrySet()) {
+            val key = entry.getKey();
+            val transactionData = entry.getValue();
+            Preconditions.checkState(null != transactionData.getKey());
+
+            // See if this entry was modified in this transaction.
+            if (transactionData.getVersion() == txn.getVersion()) {
+                modifiedKeys.add(key);
+                transactionData.setPersisted(false);
+                modifiedValues.add(transactionData);
             }
+            // make sure none of the keys used in this transaction have changed.
+            val dataFromBuffer = bufferedTxnData.get(key);
+            if (null != dataFromBuffer) {
+                if (!dataFromBuffer.isPinned()) {
+                    Preconditions.checkState(null != dataFromBuffer.getDbObject());
+                }
+                if (dataFromBuffer.getVersion() > transactionData.getVersion()) {
+                    throw new CompletionException(new StorageMetadataVersionMismatchException(
```

Comment:
You can either declare this as `throws` or use `@SneakyThrows(StorageMetadataVersionMismatchException.class)` on the method to avoid having do to this ugly wrapping.